Rework capture options dialogue

florian-kuebler opened the pull request:
This groups together some changes to the
capture options dialogue to streamline
the UI and improve UX.

1. Group each category into group boxes. This
   unifies the complete dialogue.
2. Wrap the capture options into a scroll area,
   as 1. adds quite some vertical space.
3. Change the drop down menues to radio buttons.
   This provides more space for explanations and
   allows to tie the "max raw stack" entry to the
   frame pointer unwinding method.
4. Also disable the labels and not only the edits,
   when unchecking the checkboxes/radio buttons
   for sampling, frame pointer unwinding and
   the memory tracing.
5. For the "max raw stack" edit, we now use a
   spin box that also validates that the value
   is a multiple of 8.
6. Add proper tooltips to the frame pointer case.

Screenshots (now):

1. View with scrollbar:
   http://screenshot/8WYUSNGHfMmW2Pb
2. Complete (extended window) with --devmode:
   http://screenshot/6ZGSanoG2cM6rig
3. Complete (extended window) without --devmode
   http://screenshot/8MTWsza5cK3ofgb
4. Disabled label:
   http://screenshot/KPBQYLnSr6Syr4o
5. Frame-pointer tooltip:
   http://screenshot/7pL77FgBf29SCqc

Screenshots (before):

1. With --devmode
   http://screenshot/BM8FU8SaMm8EiHE
2. Wihtout --devmode
   http://screenshot/BPiJ8YAkp5nzSLn

Test: Open capture options with/without --devmode.

antonrohr commented on src/OrbitQt/CaptureOptionsDialog.cpp:
```diff
@@ -30,56 +30,67 @@ CaptureOptionsDialog::CaptureOptionsDialog(QWidget* parent)
-  ui_->memoryWarningThresholdKbLineEdit->setValidator(&uint64_validator_);
-  ui_->maxCopyRawStackSizeLineEdit->setValidator(
-      new UInt64Validator(0, kMaxCopyRawStackSizeMaxValue, ui_->maxCopyRawStackSizeLineEdit));
+  ui_->framePointerUnwindingRadioButton->setToolTip(
+      "Frame-pointer unwinding has lower overhead, but requires all\n"
+      "binaries to be compiled with '-fno-omit-frame-pointer'.\n"
+      "Orbit also supports compiling with '-momit-leaf-frame-pointer'\n"
+      "in addition to '-fno-omit-frame-pointer', but this is optional.\n\n"
+      "Callstack samples containing functions not compiled with frame\n"
+      "pointers will result in unwinding errors.");
```
pers pref: These long tooltip texts in constants at the beginning of the file or into the .ui file

antonrohr reviewed (approved):
LGTM!

beckerhe commented on src/OrbitQt/CaptureOptionsDialog.cpp:
```diff
@@ -30,56 +30,67 @@ CaptureOptionsDialog::CaptureOptionsDialog(QWidget* parent)
+                   ui_->maxCopyRawStackSizeWidget,
+                   [this](bool checked) { ui_->maxCopyRawStackSizeWidget->setEnabled(checked); });
+  QObject::connect(ui_->collectMemoryInfoCheckBox, qOverload<bool>(&QCheckBox::toggled),
+                   ui_->memoryHorizontalLayout, [this](bool checked) {
```
Nit: To be on the safe side `ui_->memoryHorizontalLayout` should be replaced by `this`. `ui_->memorySamplingPeriodMsLabel` and `ui_->memorySamplingPeriodMsLineEdit`, etc. are child elements of the `CaptureOptionsDialog`. So this connection should get disconnected as soon as the CaptureOptionsDialog gets into deconstruction.

beckerhe commented on src/OrbitQt/CaptureOptionsDialog.cpp:
```diff
@@ -30,56 +30,67 @@ CaptureOptionsDialog::CaptureOptionsDialog(QWidget* parent)
-                     }
+
+  QObject::connect(ui_->samplingCheckBox, qOverload<bool>(&QCheckBox::toggled),
+                   ui_->samplingPeriodMsHorizontalLayout, [this](bool checked) {
```
Same as above. If you access multiple ui elements, use `this` as the receiver.

beckerhe reviewed (approved):
LGTM

dpallotti commented:
Some end-to-end tests set some capture options. I would expect some to fail now, for example the ones that change the dynamic instrumentation method.

dpallotti commented:
> * View with scrollbar:
>   http://screenshot/8WYUSNGHfMmW2Pb
> * Complete (extended window) with --devmode:
>   http://screenshot/6ZGSanoG2cM6rig
> * Complete (extended window) without --devmode
>   http://screenshot/8MTWsza5cK3ofgb

What's the default height of the dialogue? Is it possible that it doesn't have to show the scrollbar by default or is it unreasonably tall like that?

dpallotti commented on src/OrbitQt/CaptureOptionsDialog.ui:
```diff
@@ -20,411 +20,450 @@
-              <width>40</width>
-              <height>20</height>
-             </size>
+           <widget class="QGroupBox" name="unwindingMethodGroupBox">
```
I don't have an alternative, but I don't really like the nested group too much. Maybe it's only me but maybe someone has an idea on this.

dpallotti replied:
It's probably only me...

dpallotti commented on src/OrbitQt/CaptureOptionsDialog.cpp:
```diff
@@ -30,56 +30,67 @@ CaptureOptionsDialog::CaptureOptionsDialog(QWidget* parent)
-  ui_->maxCopyRawStackSizeLineEdit->setValidator(
-      new UInt64Validator(0, kMaxCopyRawStackSizeMaxValue, ui_->maxCopyRawStackSizeLineEdit));
+  ui_->framePointerUnwindingRadioButton->setToolTip(
+      "Frame-pointer unwinding has lower overhead, but requires all\n"
+      "binaries to be compiled with '-fno-omit-frame-pointer'.\n"
```
> The other tooltips (again, in the .ui file) don't force a certain column width with extra newline characters.

Discussed offline: the other tooltips are much shorter, while this one would cause a very wide tooltip.

florian-kuebler commented:
> > * View with scrollbar:
> >   http://screenshot/8WYUSNGHfMmW2Pb
> 
> > * Complete (extended window) without --devmode
> >   http://screenshot/8MTWsza5cK3ofgb
> 
> What's the default height of the dialogue? Is it possible that it doesn't have to show the scrollbar by default or is it unreasonably tall like that?

I changed the default height, such that it is not showing the scrollbar by default.

florian-kuebler commented:
> Some end-to-end tests set some capture options. I would expect some to fail now, for example the ones that change the dynamic instrumentation method.

I adjusted the E2E tests. Also added "AccessibleName" attributes to all then handles, such that we don't rely on the text.

dpallotti commented on src/OrbitQt/CaptureOptionsDialog.ui:
```diff
@@ -20,411 +20,450 @@
+           <widget class="QCheckBox" name="gpuSubmissionsCheckBox">
             <property name="text">
-             <string>3</string>
+             <string>Trace GPU queue submissions</string>
```
As discussed, I had forgotten what this check box was actually referring to.

dpallotti commented on src/OrbitQt/CaptureOptionsDialog.ui:
```diff
@@ -20,433 +20,561 @@
+       <item>
+        <widget class="QGroupBox" name="dynamicInstrumentationGroupBox">
+         <property name="toolTip">
+          <string>Method 'Orbit' has lower overhead but is still experimental.</string>
```
What do you think of using bold for "Orbit" instead of the single quotes here?
